libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp

```diff
@@ -183,10 +183,16 @@ class flystel_prime_field_gadget : public gadget<libff::Fr<ppT>>
     void generate_r1cs_witness();
 };
 
-// get the MDS matrix from the number of columns 2,3 or 4
-template<typename FieldT, size_t NumStateColumns_L>
-std::array<std::array<FieldT, NumStateColumns_L>, NumStateColumns_L>
-anemoi_permutation_mds(const FieldT g);
+// get the MDS matrix for each allowed dimension: 2,3 or 4
+template<typename ppT>
+std::array<std::array<libff::Fr<ppT>, 2>, 2> anemoi_permutation_mds_2x2(
+    const libff::Fr<ppT> g);
+template<typename ppT>
+std::array<std::array<libff::Fr<ppT>, 3>, 3> anemoi_permutation_mds_3x3(
+    const libff::Fr<ppT> g);
+template<typename ppT>
+std::array<std::array<libff::Fr<ppT>, 4>, 4> anemoi_permutation_mds_4x4(
+    const libff::Fr<ppT> g);
```

Comment:
These all have different names, so the code that uses them must still have `if-else` clauses based on the matrix size.  The idea behind the comment https://github.com/clearmatics/libsnark/pull/65#discussion_r992467709 was to use the size as a template parameter, so the calling code can just call `anemoi_permutation_mds<ppT, NumStateColumnsL>()`.

Note, as discussed w.r.t. `ppT`, `anemoi_parameters` etc, we can't template functions in precisely this way, so we template classes:
```
template<typename ppT, size_t NumStateColumns_L> class anemoi_permutation_mds;
```
where the class has a static of the form `permutation_mds()`, and we can give an implementation for `template<typename ppT> class anemoi_permutation_mds<ppT, 2>`, `template<typename ppT> class anemoi_permutation_mds<ppT, 3>`, `template<typename ppT> class anemoi_permutation_mds<ppT, 4>`.

Calling code can then use its `NumStateColumns_L` parameter and avoid any complex if statements by calling `anemoi_permutation_mds<ppT, NumStateColumnsL>::permutation_mds()`.  If calling code tries to use an invalid `NumStateColumns_L` param, it will get an error at compile (or, more precisely at link) time.  Note, we will probably also want to define the matrix types as part of the class.  This is the pattern that's used in many places in this code base, including `aemoi_parameters`.

Reply:
Understood. Fixed.